lib/spack/spack/paths.py

```diff
@@ -39,39 +134,65 @@
 operating_system_path = os.path.join(module_path, "operating_systems")
 test_path = os.path.join(module_path, "test")
 hooks_path = os.path.join(module_path, "hooks")
-opt_path = os.path.join(prefix, "opt")
 share_path = os.path.join(prefix, "share", "spack")
 etc_path = os.path.join(prefix, "etc", "spack")
-
-#
-# Things in $spack/etc/spack
-#
 default_license_dir = os.path.join(etc_path, "licenses")
-
-#
-# Things in $spack/var/spack
-#
 var_path = os.path.join(prefix, "var", "spack")
 
-# read-only things in $spack/var/spack
+
+spack_instance_id = lambda: hash.b32_hash(prefix)[:7]
+
+#: transient caches for Spack data (virtual cache, patch sha256 lookup, etc.)
+default_misc_cache_path = os.path.join(spack_xdg_state_home(), spack_instance_id(), "cache")
+
+#: concretization cache for Spack concretizations
+default_conc_cache_path = os.path.join(default_misc_cache_path, "concretization")
+
+modules_base = None
+for module_dir in ["lmod", "modules"]:
+    if dir_is_occupied(os.path.join(share_path, module_dir)):
+        modules_base = share_path
+if not modules_base:
+    modules_base = os.path.join(spack_xdg_data_home(), "modules")
+
+
+def default_install_location():
+    # Precedence for installs:
+    # 1. config:install_tree:root
+    # 2. explicitly defined SPACK_DATA_HOME
+    # 3. occupied old install path (inside spack prefix)
+    # 4. explicitly defined XDG_DATA_HOME
+    # 5. inside spack prefix (slightly different compared to old install path)
+    old_install_path = os.path.join(prefix, "opt", "spack")
+    if spack_data_home_varname in os.environ:
+        return os.path.join(spack_data_home(), "installs")
+    elif dir_is_occupied(old_install_path):
+        return old_install_path
+    else:
+        return os.path.join(spack_data_home(), "installs")
+
+
+# Environments follow the same precedence rules as installs
```

Comment:
Same as above. Also, environments already have a config to relocate them: `config:environments_root`. So we could probably embed the logic in a function for both.